pkg/codegen/customresourcegenerator.go

```diff
@@ -46,45 +54,189 @@ type CustomResourceGenerator struct {
-		if foundVersionInfos {
-			for _, version := range versionInfos {
-				name, _, _ := unstructured.NestedString(version, "name")
-				if schema, foundSchema, _ := unstructured.NestedMap(version, "schema", "openAPIV3Schema"); foundSchema {
-					schemas[name] = schema
-				}
-			}
+
+		currSpec.Items.Schema = &s
+
+		if len(currSpec.Items.Schema.Properties) == 0 {
+			return currSpec, nil
+		}
+
+		sw.Definitions[nestedDefinitionName] = s
+
+		// Reset the property to be a reference to the nested object if the item is an object.
+		refName := definitionPrefix + nestedDefinitionName
+		ref, err := jsonreference.New(refName)
+		if err != nil {
+			return currSpec, fmt.Errorf("error creating OpenAPI json reference for nested object: %w", err)
 		}
+
+		currSpec.Items.Schema.Ref = spec.Ref{Ref: ref}
+		currSpec.Items.Schema.Type = nil
+		currSpec.Items.Schema.Properties = nil
+
+		return currSpec, nil
 	}
 
-	kind, foundKind, _ := unstructured.NestedString(crd.Object, "spec", "names", "kind")
-	if !foundKind {
-		return CustomResourceGenerator{}, fmt.Errorf("could not find `spec.names.kind` field in the CRD")
+	// Recurse through the properties of the object.
+	for nestedPropertyName, nestedProperty := range currSpec.Properties {
+		// VistoriaMetrics has some weird fields - likely a typegen issue on their end, so let's skip them.
+		if nestedPropertyName == "-" {
+			delete(currSpec.Properties, nestedPropertyName)
+			continue
+		}
+		// Create a new definition for the nested object by joining the parent definition name and the property name.
+		// This is to ensure that the nested object is unique and does not conflict with other definitions.
+		nestedDefinitionName := parentName + sanitizeReferenceName(nestedPropertyName)
+
+		s, err := flattenRecursively(sw, nestedDefinitionName, nestedProperty)
+		if err != nil {
+			return currSpec, fmt.Errorf("error flattening OpenAPI spec: %w", err)
+		}
+
+		// The nested property is not an object, so we can skip adding it to the definitions and creating a reference.
+		// We check this here as we want our recursive function to inspect both arrays and objects.
+		if len(nestedProperty.Properties) == 0 {
+			continue
+		}
+
+		sw.Definitions[nestedDefinitionName] = s
+
+		// Reset the property to be a reference to the nested object.
+		refName := definitionPrefix + nestedDefinitionName
+		ref, err := jsonreference.New(refName)
+		if err != nil {
+			return currSpec, fmt.Errorf("error creating OpenAPI json reference for nested object: %w", err)
+		}
+
+		currSpec.Properties[nestedPropertyName] = spec.Schema{
+			SchemaProps: spec.SchemaProps{
+				Ref: spec.Ref{
+					Ref: ref,
+				},
+			},
+		}
 	}
-	plural, foundPlural, _ := unstructured.NestedString(crd.Object, "spec", "names", "plural")
-	if !foundPlural {
-		return CustomResourceGenerator{}, fmt.Errorf("could not find `spec.names.plural` field in the CRD")
+
+	return currSpec, nil
+}
+
+func sanitizeReferenceName(fieldName string) string {
```

Comment:
Is this basically responsible for ensuring the ref name is a valid token in all of our target languages?

Reply:
Yes! Previously, in crd2pulumi, we need to do this in a number of locations which resulted in harder to maintain code, and the bugs we saw. For package name sanitization, that resides in the p-k schemagen logic.